src/transform/src/column_knowledge.rs

```diff
@@ -106,13 +110,50 @@ impl ColumnKnowledge {
                     }
                     Ok(body_knowledge)
                 }
-                MirRelationExpr::LetRec {
-                    ids: _,
-                    values: _,
-                    body: _,
-                } => {
-                    // TODO
-                    Err(crate::TransformError::LetRecUnsupported)?
+                MirRelationExpr::LetRec { ids, values, body } => {
+                    // As a first approximation, we treat LetRec blocks as
+                    // barriers for this optimization.
+
+                    // A map to hold knowledge from shadowed bindings
+                    let mut shadowed_knowledge = BTreeMap::new();
+
+                    // We set knowledge[i][j] = DatumKnowledge::empty() for each
+                    // column j and CTE i.
+                    for (id, value) in zip_eq(ids.iter(), values.iter()) {
+                        let id = mz_expr::Id::Local(id.clone());
+                        let knowledge_new = vec![DatumKnowledge::empty(); value.arity()];
+                        if let Some(knowledge_old) = knowledge.insert(id, knowledge_new) {
+                            shadowed_knowledge.insert(id, knowledge_old);
+                        };
+                    }
+
+                    // For the first loop, we sequentially absorb into
+                    // knowledge[i][j] the result of descending into values[i].
+                    //
+                    // This is consistent with the WMR evaluation semantics of:
+                    // 1. sequential updates, and
+                    // 2. at least one iteration.
+                    for (id, value) in zip_eq(ids.iter(), values.iter_mut()) {
+                        let id = mz_expr::Id::Local(id.clone());
+                        let next_knowledge = self.harvest(value, knowledge, knowledge_stack)?;
+                        let prev_knowledge = knowledge.get_mut(&id).unwrap();
+                        for (prev, next) in zip_eq(prev_knowledge.iter_mut(), next_knowledge) {
+                            prev.absorb(&next) // Use absorb for the first iteration.
```

Comment:
Another way to think about the situation here is that
1. iterations can move us up (`union`) in the lattice of `DatumKnowledge`, and
2. the final result of our `ColumnKnowledge` implementation has to be an _upper_ bound on the ideal result (i.e., it's ok to say nullable on stuff that is actually non-nullable, but it's not ok to say non-nullable on stuff that is actually nullable).

So 1. and 2. together means that doing just one iteration instead of fixpointing would get us incorrect answers, and not just non-ideal answers.

Reply:
If I understand `DatumKnowledge` correctly, the fixpoint here should fall into the "quickly converges" category: you either know the constant value or you don't; you either are nullable or are not... so each definition has up to two possible moves up in the lattice (`(None, true)` -> learn nullability or value -> learn the other thing); with n definitions, your worst case is 2n runs of the fixpoint.

Reply:
But each definition also has multiple columns. So I think it might take as many runs to learn the nullability as the total number of columns in all the definitions.

And for the `value` component of `DatumKnowledge`, that will just start as `None`, and stay `None` always, I think.

Reply:
@ggevay / @mgree see my line of comments above that [end with this comment](https://github.com/MaterializeInc/materialize/pull/18323#discussion_r1148337500). I believe that with the proposed changes the analysis will be correct.

What is mildly concerning is that I cannot come up with an example that invalidates the current state that I can use as a regression test. I am thinking of something along the lines of

```sql
WITH MUTUALLY RECURSIVE
  r(x INT) as (
    /* populate this branch with known `x IS NOT NULL` in odd iterations */
    UNION /* select the top branch if `x IS NULL` and the bottom otherwise */
    /* populate this branch with known `x IS NULL` in even iterations */
  )
SELECT * FROM r
```

Reply:
I think the current code always removes the IS NOT NULL from the `Get l0`, if `l0` refers to itself. So it should be enough to
- have an IS NOT NULL on the `Get l0`,
- but it is invalid to remove it.

And then the current code would wrongly remove it.